### PR TITLE
[pkcs12] new package for PKCS#12 utilities

### PR DESCRIPTION
**Pull request by quinlanj**
# Why
Creates a new pkcs12 utility package meant to replace [PKCS12Utils.ts](https://github.com/expo/expo-cli/blob/master/packages/xdl/src/detach/PKCS12Utils.ts) in xdl. This is because:
- `www` depends on `xdl`, but only uses this file
- `eas-cli` has a duplicated PKCS12Utils file, and we can refactor it out to a common package
- `xdl` and `expo-cli` can eventually be refactored to use this common package

# Integration Plan
1.  🚨 🚨🚨 **Replace in `www`. This is the most urgent usecase because we need keystore pkcs12 support ASAP. This is also a good place to start because if there are any bugs, it will be the easiest to detect in Google Cloud Logs and the easiest to rollback.** 🚨🚨🚨
2. Replace in `eas-cli`, `xdl` and `expo-cli`

# How
- provides ways to extract conventional pkcs#12 files as well as keystore formatted pkcs#12 files (generated by the java `keytool` binary)
- more comprehensive tests

# Notable Changes
Functions have been changed to be more modular to account for the different p12 formats
- `getPKCS` deserializes encoded p12 file
- `getx509Certificate` and `getx509CertificateByFriendlyName` extract the certificate from a p12 object. For use by conventional p12 and keystore formatted p12 formats, respectively.
- `getCertificateFingerprint ` calculates the various hashes for a certificate object
- `getFormattedSerialNumber` calculates the serial number for a certificate object

### CODE REVIEW

File: packages/pkcs12/README.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+# Examples
+
+## Extracting a certificate from a conventional PKCS#12 file
+
+```
+const p12 = getPKCS12(base64EncodedP12, password); // deserializes encodedP12
+const certificate = getX509Certificate(p12); // extracts single certificate from p12
+const sha1Fingerprint = getCertificateFingerprint(certificate, {
+      upperCase: true,
+      hashAlgorithm: 'sha1',
```
Comment (ide) [inline packages/pkcs12/README.md:14]:
```suggestion
  upperCase: true,
  hashAlgorithm: 'sha1',
```

File: packages/pkcs12/README.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+}); // Hash like 02EC75A7181C575757BAA931FE3105B7125FF10A
+```
+
+## Extracting a certificate from a keystore formatted PKCS#12 file
```
Comment (ide) [inline packages/pkcs12/README.md:18]:
```suggestion
## Extracting a certificate from a keystore in a PKCS#12 file
```
(PKCS#12 files aren't keystore-formatted, keystores are PKCS#12-formatted)

File: packages/pkcs12/README.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+```
+const p12 = getPKCS12(base64EncodedP12, password); // deserializes encodedP12
+const certificate = getX509CertificateByFriendlyName(p12, alias); // extracts single certificate stored under alias in p12
+const sha1Fingerprint = getCertificateFingerprint(certificate, {
+      upperCase: true,
+      hashAlgorithm: 'sha1',
```
Comment (ide) [inline packages/pkcs12/README.md:25]:
```suggestion
  upperCase: true,
  hashAlgorithm: 'sha1',
```

File: packages/pkcs12/src/__tests__/pkcs12-test.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
+import {
```
Comment (ide) [inline packages/pkcs12/src/__tests__/pkcs12-test.ts:1]: name this file index-test.ts or rename the source file to pkcs12.ts

File: packages/pkcs12/src/__tests__/pkcs12-test.ts [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+  password: 'test',
+  serialNumber: '1',
+  md5Fingerprint: '2557D902372F89802767249FC6D65113',
+  sha1Fingerprint: '02EC75A7181C575757BAA931FE3105B7125FF10A',
+  sha256Fingerprint: 'FBA434E1877FBBF48CD041389EF9E1C0B511D1426C69D6D110E7D2A7DB6B5890',
+};
+
+// Generated by: keytool -genkey -alias test-alias -keystore keystore.p12 -storetype PKCS12 -keyalg RSA -storepass password -validity 730 -keysize 2048
```
Comment (ide) [inline packages/pkcs12/src/__tests__/pkcs12-test.ts:22]: This documentation is great

File: packages/pkcs12/src/__tests__/pkcs12-test.ts [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+  password: 'password',
+  alias: 'test-alias',
+  md5Fingerprint: 'CCD9D0FD20A862B1E67EA838C99DD0C5',
+  sha1Fingerprint: '2EF9743A1BF12189AF5503787B54BFD3254B46A8',
+  sha256Fingerprint: 'EECDD35E90BB272D52EC261B4063C5D586980795452340525F9C8CC56DB53E2B',
+};
+describe('PKCS12 Tests', () => {
+  it('Conventional P12 Cert Fingerprints', async () => {
```
Comment (ide) [inline packages/pkcs12/src/__tests__/pkcs12-test.ts:35]: Write test descriptions to fit in the sentence, "It ...":
```suggestion
  it('computes fingerprints of certificates in conventional p12 files', async () => {
```

File: packages/pkcs12/src/__tests__/pkcs12-test.ts [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    });
+    const sha256Fingerprint = getCertificateFingerprint(certificate, {
+      upperCase: true,
+      hashAlgorithm: 'sha256',
+    });
+    expect(md5Fingerprint).toEqual(expectedMd5Fingerprint);
+    expect(sha1Fingerprint).toEqual(expectedSha1Fingerprint);
+    expect(sha256Fingerprint).toEqual(expectedSha256Fingerprint);
+  });
+  it('Keystore P12 Cert Fingerprints', async () => {
```
Comment (ide) [inline packages/pkcs12/src/__tests__/pkcs12-test.ts:61]:
```suggestion
  it('computes fingerprints of certificates in p12 keystores', async () => {
```

File: packages/pkcs12/src/__tests__/pkcs12-test.ts [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    });
+    const sha256Fingerprint = getCertificateFingerprint(certificate, {
+      upperCase: true,
+      hashAlgorithm: 'sha256',
+    });
+    expect(md5Fingerprint).toEqual(expectedMd5Fingerprint);
+    expect(sha1Fingerprint).toEqual(expectedSha1Fingerprint);
+    expect(sha256Fingerprint).toEqual(expectedSha256Fingerprint);
+  });
+  it('Conventional P12 Serial Number', async () => {
```
Comment (ide) [inline packages/pkcs12/src/__tests__/pkcs12-test.ts:88]:
```suggestion
  it('reads X.509 certificate serial numbers from conventional p12 files', async () => {
```

File: packages/pkcs12/src/__tests__/pkcs12-test.ts [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    const { base64EncodedP12, password, serialNumber: expectedSerialNumber } = conventionalP12;
+    const p12 = getPKCS12(base64EncodedP12, password);
+    const certificate = getX509Certificate(p12);
+    const serialNumber = getFormattedSerialNumber(certificate);
+    expect(serialNumber).toEqual(expectedSerialNumber);
+  });
+  it('Conventional P12 Certificate Data', async () => {
```
Comment (ide) [inline packages/pkcs12/src/__tests__/pkcs12-test.ts:95]:
```suggestion
  it('reads X.509 certificates from conventional p12 files', async () => {
```

File: packages/pkcs12/src/__tests__/pkcs12-test.ts [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+  password: 'password',
+  alias: 'test-alias',
+  md5Fingerprint: 'CCD9D0FD20A862B1E67EA838C99DD0C5',
+  sha1Fingerprint: '2EF9743A1BF12189AF5503787B54BFD3254B46A8',
+  sha256Fingerprint: 'EECDD35E90BB272D52EC261B4063C5D586980795452340525F9C8CC56DB53E2B',
+};
+describe('PKCS12 Tests', () => {
```
Comment (ide) [inline packages/pkcs12/src/__tests__/pkcs12-test.ts:34]:
```suggestion
describe('reading PKCS#12 files', () => {
```
(We're not using the tests to describe the behavior of the tests; we're describing the behavior of the library.)

File: packages/pkcs12/src/__tests__/pkcs12-test.ts [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+    const serialNumber = getFormattedSerialNumber(certificate);
+    expect(serialNumber).toEqual(expectedSerialNumber);
+  });
+  it('Conventional P12 Certificate Data', async () => {
+    const { base64EncodedP12, password } = conventionalP12;
+    const p12 = getPKCS12(base64EncodedP12, password);
+    const certificate = getX509Certificate(p12);
+    expect(certificate).toMatchSnapshot();
+  });
+});
```
Comment (ide) [inline packages/pkcs12/src/__tests__/pkcs12-test.ts:101]: Consider adding tests for when getX509Certificate is called on a keystore, and when getX509CertificateByFriendlyName is called on a conventional .p12 file.

File: packages/pkcs12/src/index.ts [hunk 12]
```diff
@@ -0,0 +1,106 @@
+import crypto, { HexBase64Latin1Encoding, Utf8AsciiLatin1Encoding } from 'crypto';
+import forge from 'node-forge';
+
+/**
+ * returns serial number as an uppercased hexidecimal string
```
Comment (ide) [inline packages/pkcs12/src/index.ts:5]:
```suggestion
 * Returns the serial number of the given X.509 certificate as an uppercased hexadecimal string
```

File: packages/pkcs12/src/index.ts [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+  const certificate = bags[0].cert;
+  if (!certificate) {
+    throw new Error('PKCS12: bag is not a certificate');
+  }
+  return certificate;
+}
+
+/**
+ * Extracts a certificate from PKCS#12
+ * This is assumed to be a PKCS#12 in Keystore format where the friendlyName (alias) contains a PrivateKeyEntry
```
Comment (ide) [inline packages/pkcs12/src/index.ts:31]:
```suggestion
 * This is assumed to be a PKCS#12 containing a keystore where the friendlyName (alias) contains a PrivateKeyEntry
```

File: packages/pkcs12/src/index.ts [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+export function getPKCS12(
+  p12BufferOrBase64String: Buffer | string,
+  maybePassword: string | null
+): forge.pkcs12.Pkcs12Pfx {
+  const base64EncodedP12 = Buffer.isBuffer(p12BufferOrBase64String)
+    ? p12BufferOrBase64String.toString('base64')
+    : p12BufferOrBase64String;
+  const password = String(maybePassword || '');
```
Comment (ide) [inline packages/pkcs12/src/index.ts:58]:
```suggestion
  const password = String(maybePassword ?? '');
```

File: packages/pkcs12/src/index.ts [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+    hashAlgorithm?: string;
+    hashEncoding?: HexBase64Latin1Encoding;
+    inputEncoding?: Utf8AsciiLatin1Encoding;
+  }
+): string {
+  const hash = crypto.createHash(hashAlgorithm ?? 'sha1');
+  if (inputEncoding) {
+    hash.update(data, inputEncoding);
+  } else {
+    hash.update(data); // use node default inputEncoding
```
Comment (ide) [inline packages/pkcs12/src/index.ts:82]:
```suggestion
    hash.update(data); // use Node's default inputEncoding
```

File: packages/pkcs12/src/index.ts [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+  }
+  const digest = hash.digest(hashEncoding ?? 'hex');
+  return upperCase ? digest.toUpperCase() : digest;
```
Comment (ide) [inline packages/pkcs12/src/index.ts:85]: tbh I'd leave off this upperCase option and just make the caller responsible for doing this if they want it.

**Review by ide — CHANGES_REQUESTED** [review]

File: packages/pkcs12/src/index.ts [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+    throw new Error(`PKCS12: No certificates found under friendlyName: ${friendlyName}`);
+  }
+  const certificate = bags[0].cert;
+  if (!certificate) {
+    throw new Error('PKCS12: bag is not a certificate');
+  }
+  return certificate;
+}
+
+export function getPKCS12(
```
Comment (wkozyra95) [inline packages/pkcs12/src/index.ts:51]: maybe `parse` or `decode` would be better here

**Review by wkozyra95 — APPROVED** [review]

File: packages/pkcs12/src/index.ts (hunk 17 above)
Comment (fson) [reply packages/pkcs12/src/index.ts:51]: Decrypt?

Comment (wkozyra95) [reply packages/pkcs12/src/index.ts:51]: I'm not sure `pkcs12FromAsn1` returns decrypted values, but if that is the case then yes decrypt would be a better name.

File: packages/pkcs12/src/index.ts [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+  p12BufferOrBase64String: Buffer | string,
+  maybePassword: string | null
+): forge.pkcs12.Pkcs12Pfx {
+  const base64EncodedP12 = Buffer.isBuffer(p12BufferOrBase64String)
+    ? p12BufferOrBase64String.toString('base64')
+    : p12BufferOrBase64String;
+  const password = String(maybePassword || '');
+  const p12Der = forge.util.decode64(base64EncodedP12);
```
Comment (ide) [inline packages/pkcs12/src/index.ts:59]: what is the format of p12Der? just want to make sure we're not doing extra work converting a buffer to base64 and then converting it back to some buffer-like data type.

File: packages/pkcs12/src/index.ts (hunk 17 above)
Comment (quinlanj) [reply packages/pkcs12/src/index.ts:51]: i think parse would be better here -- it definitely parse/decodes the base64 p12 to an object. It also optionally uses the password to decrypt the p12 if the content type is marked as 'encryptedData' https://github.com/digitalbazaar/forge/blob/588c41062d9a13f8dc91be3723b159c6cc434b15/lib/pkcs12.js#L554

File: packages/pkcs12/src/index.ts (hunk 18 above)
Comment (quinlanj) [reply packages/pkcs12/src/index.ts:59]: p12Der is a binary encoded string. The way we are getting the p12 file is how `node-forge` recommends in their [docs](https://github.com/digitalbazaar/forge#pkcs12). The methods they expose are basically to get the base64 p12 -> decode it into a binary encoded string (p12Der) -> convert it to an asn1 object -> convert it to a pkcs12 object.

**Review by ide — APPROVED** [review]

**Pull request merged**